### PR TITLE
Fix Truffle API deprecations

### PR DESCRIPTION
**Pull request by chrisseaton**

### CODE REVIEW

**Comment by chrisseaton** [comment]
Where we're using `FIRST_INSTANCE` of the `RubyContext` I need to ask Christian Humer how he intends us to fix that - I think he has some new API to do it.

File: src/test/java/org/truffleruby/RubyTest.java [hunk 1]
```diff
@@ -39,20 +39,20 @@
...
     }
 
-    protected void testInEngine(Consumer<RubyContext> test) {
+    protected void testInEngine(Runnable test) {
         final PolyglotEngine engine = PolyglotEngine.newBuilder()
                 .config(RubyLanguage.MIME_TYPE, OptionsCatalog.EXCEPTIONS_TRANSLATE_ASSERT.getName(), false)
                 .config(RubyLanguage.MIME_TYPE, OptionsCatalog.BASICOPS_INLINE.getName(), false)
                 .globalSymbol("action", JavaInterop.asTruffleFunction(Runnable.class, () ->
-                    test.accept(RubyContext.getInstance())
+                    test.run()
```
Comment (eregon) [inline src/test/java/org/truffleruby/RubyTest.java]: `test` could be passed directly without an extra lambda.

File: src/main/java/org/truffleruby/language/LazyRubyRootNode.java [hunk 2]
```diff
@@ -34,32 +35,35 @@
...
     }
 
     @Override
     public Object execute(VirtualFrame frame) {
-        final RubyContext context = RubyLanguage.INSTANCE.unprotectedFindContext(findContextNode);
+        final RubyContext context = contextReference.get();
 
         if (cachedContext == null) {
             CompilerDirectives.transferToInterpreterAndInvalidate();
+            context.ensureSetupFinished();
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/LazyRubyRootNode.java]: This is a bit ugly. Is there a better way @chumer ?

Comment (chrisseaton) [inline src/main/java/org/truffleruby/language/LazyRubyRootNode.java]: I don't think he'll know what you are asking without any context.

In the new APIs, you cannot find the current context until the constructor of the context has finished. We run code during context creation. Therefore I've made the parts of context creation that need the context lazy, done when a lazy root node is first resolved.

File: src/main/java/org/truffleruby/RubyContext.java [hunk 3]
```diff
@@ -167,10 +174,7 @@ public RubyContext(TruffleLanguage.Env env) {
 
             // Capture known builtin methods
 
-            final Instrumenter instrumenter = env.lookup(Instrumenter.class);
-            traceManager = new TraceManager(this, instrumenter);
             coreMethods = new CoreMethods(this);
-            coverageManager = new CoverageManager(this, instrumenter);
 
             // Load the reset of the core library
```
Comment (eregon) [inline src/main/java/org/truffleruby/RubyContext.java]: Typo (already existed): reset -> rest but maybe a better description is:
```java
// Load the part of the core library defined in Ruby
```

File: src/main/java/org/truffleruby/RubyContext.java [hunk 4]
```diff
@@ -89,35 +93,31 @@
...
-
-    public static ThreadLocal<RubyContext> contextsBeingCreated = new ThreadLocal<>();
-
-    public RubyContext(TruffleLanguage.Env env) {
-        contextsBeingCreated.set(this);
-
-        try {
-            this.env = env;
+    public RubyContext(RubyLanguage language, TruffleLanguage.Env env) {
+        if (FIRST_INSTANCE == null) {
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/RubyContext.java]: Do we need to worry about races here?

Comment (chrisseaton) [inline src/main/java/org/truffleruby/RubyContext.java]: Only theoretically. It'll be fixed before it becomes a practical problem.

**Comment by chrisseaton** [comment]
@nirvdrum @eregon can you two check changes and approve. The ugly bits - the static reference and the explicit `ensureSetupFinished` call are open issues that I potentially need new APIs to fix. This PR fixes enough to unblock other critical work so I need it merged.

**Comment by eregon** [comment]
@chrisseaton Isn't TruffleLanguage#initializeContext() exactly what we want? Could you try it?
https://graalvm.github.io/graal/truffle/javadoc/com/oracle/truffle/api/TruffleLanguage.html#initializeContext-C-

**Comment by chrisseaton** [comment]
Yes that was it. Done.

**Comment by eregon** [comment]
Could we use a ThreadLocal instead of LATEST_INSTANCE (that we set in the constructor and when creating a Ruby thread and we only clear in shutdown() and when a Ruby thread ends)? That would still work correctly with multiple contexts in one process.
For the case of the test, if it needs to access the context from a non-Ruby thread, maybe we can use https://graalvm.github.io/graal/truffle/javadoc/com/oracle/truffle/api/vm/PolyglotEngine.html#findGlobalSymbol-java.lang.String-

**Comment by chrisseaton** [comment]
I don't want to spend time finding a better workaround though - @chumer says he has a proper solution coming soon in a new API next week. Again, this PR is to unstick other issues marked as critical.

**Comment by eregon** [comment]
OK, seems good enough to me as a temporary solution.
If Travis doesn't pass due to mx bugs, maybe try with truffle as a source suite as a workaround.

**Pull request merged**